Chapter 10: Controllers and Routing (#16)
LaunchCodeEducation/csharp-web-dev-curriculum: content/aspdotnet-controllers-and-routes/studio/_index.md

## Creating Controllers

In your project, add a controller class called `SkillsController` inside the `Controllers` folder.
Inside `SkillsController`, you will add several methods to accomplish the following:

1. At `localhost:5001/skills`, add text that states the three possible programming languages someone could be working on.
   You need to have an `h1` with the title "Skills Tracker", an `h2`, and an `ol` containing three programming languages 
   of your choosing.
1. At `localhost:5001/skills/form`, add a form that lets the user enter a date and choose their learning progress in each of the programming skills listed at `/skills`.

   {{% notice green "Tip" "rocket" %}}
   
   Use the [date input type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date). Note that the value is a `string` type. 
   
   For each of the skills, use `select` elements for selecting a description of progress. 

   {{% /notice %}}

1. Also at `localhost:5001/skills/form`, use `[HttpPost]` and request parameters to update the HTML with an `h1` stating the 
   submitted date and a list showing the user's learning progress in the three programming languages they chose.

## End Result

> At the end of the studio, when you navigate to `localhost:5001/skills`, you should see the following:
> 
> ![Image showing functioning home page](pictures/studio-home-page.png)
> 
>    Skill Tracker app home page.
> 
> When you navigate to `localhost:5001/skills/form`, you should see a blank form that looks something like: 
> 
> ![Image showing the blank form](pictures/blank-studio-form.png)
> 
>    Skill Tracker app blank skill form.
> 
> If you fill out the form, your page may render like so:
> 
> ![Image showing the web page with information from the completed form](pictures/completed-studio-form.png)
> 
>    Skill Tracker app result of skill form submission.

Reviewer: ```suggestion
At the end of the studio, when you navigate to `localhost:5001/skills`, you should see the following:

{{< rawhtml >}}
   <img src="pictures/studio-home-page.png" alt="Skill Tracker app home page" width="80%" />
{{< /rawhtml>}}

When you navigate to `localhost:5001/skills/form`, you should see a blank form that looks something like: 

{{< rawhtml >}}
   <img src="pictures/blank-studio-form.png" alt="Skill Tracker app blank skill form" width="80%" />
{{< /rawhtml>}}

If you fill out the form, your page may render like so:

{{< rawhtml >}}
   <img src="pictures/completed-studio-form.png" alt="Skill Tracker app result of skill form submission" width="80%" />
{{< /rawhtml>}}
```

Reviewer: The above is a suggestion based on what you did with the images in the reading.

At the end of the studio, your final project should be able to take input from a user via a form and post the information 
in a way that is easy to read.

## Create the MVC Project

In Visual Studio, create a new ASP.NET project using the *Web Application (Model-View-Controller)* option. Name your 
new project `SkillsTracker`. Once created, run the project and go to `localhost:5001` to ensure that there are no 
underlying build errors to fix before coding. Now you are ready to start tracking skills.

## Creating Controllers

In your project, add a controller class called `SkillsController` inside the `Controllers` folder.
Inside `SkillsController`, you will add several methods to accomplish the following:

1. At `localhost:5001/skills`, add text that states the three possible programming languages someone could be working on.
   You need to have an `h1` with the title "Skills Tracker", an `h2`, and an `ol` containing three programming languages 
   of your choosing.
1. At `localhost:5001/skills/form`, add a form that lets the user enter a date and choose their learning progress in each of the programming skills listed at `/skills`.

>    {{% notice green "Tip" "rocket" %}}
> 
>       Use the [date input type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date). Note that the value is 
>       a `string` type. 
>       
>       For each of the skills, use `select` elements for selecting a description of progress. 
> 
>    {{% /notice %}}

Reviewer: ```suggestion
   {{% notice green "Tip" "rocket" %}}
   
   Use the [date input type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date). Note that the value is a `string` type. 
   
   For each of the skills, use `select` elements for selecting a description of progress. 

   {{% /notice %}}
```

1. Also at `localhost:5001/skills/form`, use `[HttpPost]` and request parameters to update the HTML with an `h1` stating the 
   submitted date and a list showing the user's learning progress in the three programming languages they chose.

## End Result

At the end of the studio, when you navigate to `localhost:5001/skills`, you should see the following:

> .. figure:: figures/studio-home-page.png
>    :alt: Image showing functioning home page.

Reviewer: ```suggestion
![Image showing functioning home page](pictures/studio-home-page.png)
```

Reviewer: Need to create a pictures folder inside the studio folder and add the following:

studio-home-page.png
blank-studio-form.png
completed-studio-form.png

   Skill Tracker app home page.

When you navigate to `localhost:5001/skills/form`, you should see a blank form that looks something like: 

> .. figure:: figures/blank-studio-form.png
>    :alt: Image showing the blank form.

Reviewer: ```suggestion
![Image showing the blank form](pictures/blank-studio-form.png)
```

   Skill Tracker app blank skill form.

If you fill out the form, your page may render like so:

> .. figure:: figures/completed-studio-form.png
>    :alt: Image showing the web page with information from the completed form.

Reviewer: ```suggestion
![Image showing the web page with information from the completed form](pictures/completed-studio-form.png)
```
